### PR TITLE
Build EcoCI plugin

### LINKED ISSUE #909: Create EcoCI plugin
**Why**: Sub of #898 
**What**: Create a EcoCI plugin

We recently merged a PR adding the EcoCI tool to our CI/CD which uses a model to estimate the energy and carbon consumption of our Github Actions. The data is presented on a dashboard and also exposed via an API. We would like to be able to pull this data into an IF manifest file so that the energy consumed and carbon emitted by our CI/CD can be accounted for in more comprehensive assessments of our project's imapcts.

Scope of work:
- [x] Create plugin in a new IF repository (it won't be a builtin)
- [x] Add README docs
- [ ] List plugin on explorer.if.greensoftware.foudnation

**Acceptance criteria**

- [x] The plugin retrieves energy and carbon values from the EcoCI API
The plugin should sum all the energy and carbon values for all the individual components of the CI/CD and do any unit conversions necessary to express energy in kWh and carbon in gCO2eq. The start and end values should be grabbed from the first and last timestamps in the `input` data (with the end being the last timestep plus its duration) for the node where the plugin is invoked, or if only one timestamp exists then it should be the first timestamp plus the duration.

**Given** the plugin works as intended
**When** I run the following manifest:

```yaml
name: eco-ci demo
description: 
tags:
initialize:
  plugins:
    ci/cd:
      method: EcoCiApi
      path: "eco-ci-plugin"
      global-config:
        repo: "Green-Software-Foundation/if"
        branch: main
        workflow: "66389738"
tree:
  children:
    child:
      pipeline:
        - ci/cd
      config:
      inputs:
        - timestamp: '2024-07-10T11:51:48.467730+00:00'
          duration: 3600
```

I should receive the following output (actual values will be different):

```yaml
name: eco-ci demo
description: 
tags:
initialize:
  plugins:
    ci/cd:
      method: EcoCiApi
      path: "eco-ci-plugin"
      global-config:
        repo: "Green-Software-Foundation/if"
        branch: main
        workflow: "66389738"
tree:
  children:
    child:
      pipeline:
        - ci/cd
      config:
      inputs:
        - timestamp: '2024-07-10T11:51:48.467730+00:00'
          duration: 3600
          energy: 0.0003 #kWh`
          carbon: 0.00004 #gCo2eq
```

### PR DESCRIPTION
**Pull request by manushak**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# README
+
+## Overview
+
+The EcoCI plugin retrieves the `energy` and the `carbon` of the workflow of the a given repository using the [Green Metrics](https://metrics.green-coding.io/index.html) API, if the workflow of the repository is listed in the [CI projects](https://metrics.green-coding.io/ci-index.html).
```
Comment (narekhovhannisyan) [inline README.md]:
```suggestion
The EcoCI plugin retrieves the `energy` and the `carbon` of the workflow of the given repository using the [Green Metrics](https://metrics.green-coding.io/index.html) API, if the workflow of the repository is listed in the [CI projects](https://metrics.green-coding.io/ci-index.html).
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+## Implementation
+
+The EcoCI plugin fetches the `energy` and `carbon` of the specified workflow from the Green Metrics API for each entry in a manifest's input data.
+
+- The `timestamp` and `duration` fields in the input data are used to filter the `energy` and `carbon` data. The relevant endpoint is documented [here](https://api.green-coding.io/docs).
+- The Green Metrics API provides data if the `timestamp` have zero time – e.g., is exact date.
+
+## Usage
+
+To run the `EcoCI` plugin, an instance of `ExecutePlugin` must be created. Then, the plugin's `execute()` method can be called, passing required arguments to it.
```
Comment (narekhovhannisyan) [inline README.md]:
```suggestion
To run the `EcoCI` plugin, an instance of `ExecutePlugin` must be created. Then, the plugin's `execute()` method can be called, passing the required arguments to it.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+The plugin can throw the following errors:
+
+- `APIRequestError`: caused by a problem retrieving data from the API. The error message returned from the API is echoed in the IF error message.
+- `InputValidationError`: thrown when some data in the input is incorrect or missing.
+- `GlobalConfigError`: thrown when some data in the global config is incorrect or missing.
+
+## Integration into Impact Framework
+
+You can load the plugin directly from this EcoCI repository. Simply run `npm install -g https://github.com/Green-Software-Foundation/if-eco-ci-plugin`
```
Comment (narekhovhannisyan) [inline README.md]: I guess we have to publish it to npm

File: src/__tests__/unit/lib/eco-ci.test.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
+import axios from 'axios';
+import { ERRORS } from '@grnsft/if-core/utils';
+
+import { EcoCI } from '../../../lib/eco-ci';
+
+import { getMockRepoMetrics } from '../../../__mocks__/api';
+
+jest.mock('axios');
```
Comment (narekhovhannisyan) [inline src/__tests__/unit/lib/eco-ci.test.ts]: there is `axios-mock-adapter` library which has good toolset for testing API calls

File: README.md (hunk 3 above)
Comment (manushak) [reply README.md]: This was added by mistake, I've updated it already.

**Pull request merged**